book/src/04-non-video/01-buttons.md

```diff
@@ -3,3 +3,100 @@
 It's all well and good to just show a picture, even to show an animation, but if
 we want a game we have to let the user interact with something.
 
+## Key Input Register
+
+* KEYINPUT, `0x400_0130`, `u16`, read only
+
+This little `u16` stores the status of _all_ the buttons on the GBA, all at
+once. There's only 10 of them, and we have 16 bits to work with, so that sounds
+easy. However, there's a bit of a catch. The register follows a "low-active"
+convention, where pressing a button _clears_ that bit until it's released.
+
+```rust
+const NO_BUTTONS_PRESSED: u16 = 0b0000_0011_1111_1111;
+```
+
+The buttons are, going up in order from the 0th bit:
+
+* A
+* B
+* Select
+* Start
+* Right
+* Left
+* Up
+* Down
+* R
+* L
+
+Bits above that are not used. However, since the left and right directions, as
+well as the up and down directions, can never be pressed at the same time, the
+`KEYINPUT` register should never read as zero. Of course, the register _might_
+read as zero if someone is using an emulator that allows for such inputs, so I
+wouldn't go so far as to make it be `NonZeroU16` or anything like that.
+
+When programming, we usually are thinking of what buttons we want to have _be
+pressed_ instead of buttons we want to have _not be pressed_. This means that we
+need an inversion to happen somewhere along the line. The easiest moment of
+inversion is immediately as you read in from the register and wrap the value up
+in a newtype.
+
+```rust
+pub fn read_key_input() -> KeyInput {
+  KeyInput(KEYINPUT.read() ^ 0b0000_0011_1111_1111)
+}
+```
+
+Now the KeyInput you get can be checked for what buttons are pressed by checking
+for a set bit like you'd do anywhere else.
+
+```rust
+impl KeyInput {
+  pub fn a_pressed(self) -> bool {
+    (self.0 & A_BIT) > 0
+  }
+}
+```
+
+Note that the current `KEYINPUT` value changes in real time as the user presses
+or releases the buttons. To account for this, it's best to read the value just
+once per game frame and then use that single value as if it was the input across
+the whole frame. If you've worked with polling input before that should sound
+totally normal. If not, just remember to call `read_key_input` once per frame
+and then use that `KeyInput` value across the whole frame.
+
+### Detecting New Presses
```

Comment:
Debouncing would be a good topic to talk about here. I'm assuming the hardware doesn't debounce for the software.

Reply:
During an particular CPU cycle the buttons will be either on or off, but there's no more special processing than that.